### PR TITLE
Deprecate copy copyable

### PR DESCRIPTION
**Pull request by lupestro**
Initial implementation roughed in with a few existing issues:
* I want to export it so that copy and copyable end up exported from a single module even though the sources are originating in separate files. Not sure how to orchestrate that.
* API help isn't getting generated properly
* copy doesn't appear in API help index so markup isn't right
* I've never actually published an NPM package, so the package.json needs close scrutiny

I'm not sure what else I missed along the way...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ember-copy
 ==============================================================================
 
-[Short description of the addon.]
+This addon supplies a copy() function and Copyable mixin to replace the function and mixin now being deprecated in Ember.js.
```
Comment (rwjblue) [inline README.md]: Should we cross link the RFC here?

File: README.md [hunk 2]
```diff
@@ -14,15 +14,22 @@ ember install ember-copy
...
+
 
 
 Contributing
 ------------------------------------------------------------------------------
 
 ### Installation
 
-* `git clone <repository-url>`
+* `git clone https:\\github.com\emberjs\ember-copy`
```
Comment (rwjblue) [inline README.md]: Hmm, wrong slashes here?

File: app/ember-copy.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export { default as copy } from 'ember-copy/copy';
```
Comment (rwjblue) [inline app/ember-copy.js]: Shouldn't need this file at all (we don't want to export in the app directory)...

File: package.json [hunk 4]
```diff
@@ -16,16 +17,24 @@
...
     "lint:js": "eslint ./*.js addon addon-test-support app config lib server test-support tests",
     "start": "ember serve",
-    "test": "ember try:each"
+    "test": "ember test",
+    "test:all": "ember try:each"
   },
   "dependencies": {
-    "ember-cli-babel": "^6.6.0"
+    "ember-cli-babel": "^6.6.0",
+    "yuidocjs": "^0.10.2"
```
Comment (rwjblue) [inline package.json]: This probably doesn't need to be a dependency (seems like a devDep)?

**Review by rwjblue — COMMENTED** [review]
Overall looks really good!

File: addon/copy.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import { assert } from '@ember/debug';
+import EmberObject from '@ember/object';
+import Copyable from './copyable';
+
+/**
+ @module ember-copy
```
Comment (pzuraq) [inline addon/copy.js]: You shouldn't need to add `@module` anywhere with our version of YUIDoc, YUIDoc actually does supply filenames when you build plain JSON, so we infer the modules directly from that

File: addon/copy.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  returned if the optional `deep` argument is `true`.
+
+  If the passed object implements the `Copyable` interface, then this
+  function will delegate to the object's `copy()` method and return the
+  result. See `Copyable` for further details.
+
+  For primitive values (which are immutable in JavaScript), the passed object
+  is simply returned.
+
+  @method copy
```
Comment (pzuraq) [inline addon/copy.js]: We also have made some changes to make things a bit more ergonomic for modern Javascript modules:

* `@function` should be used to document module functions
* `@const`, `@constant`, or `@variable` should be used to document exported variables

File: addon/copyable.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+/**
+  Implements some standard methods for copying an object. Add this mixin to
+  any object you create that can create a copy of itself. This mixin is
+  added automatically to the built-in array.
+
+  You should generally implement the `copy()` method to return a copy of the
+  receiver.
+
+  @class Copyable
+  @namespace ember-copy
```
Comment (pzuraq) [inline addon/copyable.js]: I don't think we'll actually do anything with the `@namespace` property here

File: addon/copy.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  @static
```
Comment (pzuraq) [inline addon/copy.js]: You also won't need this

File: addon/copy.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  @param {Object} obj The object to clone
+  @param {Boolean} [deep=false] If true, a deep copy of the object is made.
+  @return {Object} The copied object
+  @public
```
Comment (pzuraq) [inline addon/copy.js]: Things are considered public by default, but doesn't hurt to keep this

**Review by pzuraq — COMMENTED** [review]
Looks like the docs are broken in `ec-addon-docs` directly, but we also have a slightly [modified version of YUIDoc which we're using](https://github.com/ember-learn/ember-cli-addon-docs-yuidoc). You should be able to make some small changes and get things working better, the docs for the custom YUIDoc parser definitely need to be more thorough so let me know how we can make it clearer!

**Comment by lupestro** [comment]
`copy` and `Copyable` are each in their own JS modules right now. 
I'm not sure how to re-export them so that the user can say: 
```import {copy, Copyable} from 'ember-copy';``` 
This is analogous to what the users were already doing:
```import {copy, Copyable} from '@ember/object/internal';``` 
I'm new to add-ons and I'm not sure how emberjs does its re-exports either.
I could use some advice about where to put the JS code to make this "look right". 

Aside from this and one API doc issue, I think this add-on is starting to look somewhere near ready to find a new home. With ember-data now in good shape, I'll start on the deprecation itself.

**Comment by pzuraq** [comment]
you should be able to do:

```js
// addon/index.js
export { default as copy } from './copy';
export { default as Copyable  } from './copyable';
```

**Comment by lupestro** [comment]
It's time to make this exist - we can move it later.

**Comment by rwjblue** [comment]
👏👏👏

**Pull request merged**